File: services/nuldoc/nuldoc-src/pages/AtomPage.ts
import { Feed } from "../generators/atom.ts";
import { elem, Element } from "../dom.ts";

export default function AtomPage({ feed }: { feed: Feed }): Element {
  return elem(
    "feed",
    { xmlns: "http://www.w3.org/2005/Atom" },
    elem("id", {}, feed.id),
    elem("title", {}, feed.title),
    elem("link", { rel: "alternate", href: feed.linkToAlternate }),
    elem("link", { rel: "self", href: feed.linkToSelf }),
    elem("author", {}, elem("name", {}, feed.author)),
    elem("updated", {}, feed.updated),
    ...feed.entries.map((entry) =>
      elem(
        "entry",
        {},
        elem("id", {}, entry.id),
        elem("link", { rel: "alternate", href: entry.linkToAlternate }),
        elem("title", {}, entry.title),
        elem("summary", {}, entry.summary),
        elem("published", {}, entry.published),
        elem("updated", {}, entry.updated),
      )
    ),
  );
}

File: services/nuldoc/lib/nuldoc/pages/atom_page.rb
module Nuldoc
  module Pages
    class AtomPage
      def initialize(feed:)
        @feed = feed
      end

      def render
        feed = @feed
        DOM::AtomXMLBuilder.new.build do
          feed xmlns: 'http://www.w3.org/2005/Atom' do
            id { text feed.id }
            title { text feed.title }
            link rel: 'alternate', href: feed.link_to_alternate
            link rel: 'self', href: feed.link_to_self
            author { name { text feed.author } }
            updated { text feed.updated }
            feed.entries.each do |entry|
              entry do
                id { text entry.id }
                link rel: 'alternate', href: entry.link_to_alternate
                title { text entry.title }
                summary { text entry.summary }
                published { text entry.published }
                updated { text entry.updated }
              end
            end
          end
        end
      end
    end
  end
end
